TLC5957.cpp

```diff
@@ -0,0 +1,366 @@
+            for (uint8_t led_channel_index = 0; led_channel_index < LEDS_PER_CHIP; led_channel_index++)
+                totalCurrent_channel[color_channel_index] += grayscale_data[chip][led_channel_index][color_channel_index];
+
+    double totalCurrent = 0;
+    for (uint8_t color_channel_index = 0; color_channel_index < COLOR_CHANNEL_COUNT; color_channel_index++)
+        totalCurrent += (_maxOutputCurrent * _CC[color_channel_index] / 511 * _BC)
+                        * (totalCurrent_channel[color_channel_index] / 0xFFFF);
+
+    return totalCurrent;
+}
+
+void TLC5957::setLodDetection(bool lod_bit_0, bool lod_bit_1)
+{
+    uint64_t new_data = lod_bit_0 | (lod_bit_1 << 1);
+    _function_data |= LOD_DETECTION_MASK & new_data;
+}
+
+void TLC5957::setTdSelection(bool td_bit_0, bool td_bit_1)
+{
+    uint64_t new_data = (td_bit_0 << 2) | (td_bit_1 << 3);
+    _function_data |= TD_SELECTION_MASK & new_data;
+}
+
+void TLC5957::setGroupDelaySelect(bool group_delay_bit)
+{
+    uint64_t new_data = group_delay_bit << 4;
+    _function_data |= GROUP_DELAY_SELECT_MASK & new_data;
+}
+
+void TLC5957::setRefreshMode(bool refresh_mode_bit)
+{
+    uint64_t new_data = refresh_mode_bit << 5;
+    _function_data |= REFRESH_MODE_MASK & new_data;
+}
+
+void TLC5957::setGclkEdgeSelect(bool gsclk_edge_bit)
+{
+    uint64_t new_data = gsclk_edge_bit << 6;
+    _function_data |= GSCLK_EDGE_SELECT_MASK & new_data;
+}
+
+void TLC5957::setPrechargeMode(bool precharge_bit)
+{
+    uint64_t new_data = precharge_bit << 7;
+    _function_data |= PRECHARGE_MODE_MASK & new_data;
+}
+
+void TLC5957::setEspwm(bool espwm_bit)
+{
+    uint64_t new_data = espwm_bit << 8;
+    _function_data |= ESPWM_MASK & new_data;
+}
+
+void TLC5957::setBlueCompensation(bool blue_compensation_bit)
+{
+    uint64_t new_data = blue_compensation_bit << 9;
+    _function_data |= BLUE_COMPENSATION_MASK & new_data;
+}
+
+void TLC5957::setSclkEdgeSelect(bool sclk_edge_bit)
+{
+    uint64_t new_data = sclk_edge_bit << 10;
+    _function_data |= SCLK_EDGE_SELECT & new_data;
+}
+
+void TLC5957::setLowGsEnhancement(bool low_gs_bit_0, bool low_gs_bit_1, bool low_gs_bit_3)
+{
+    uint64_t new_data = (low_gs_bit_0 << 11) | (low_gs_bit_1 << 12) | (low_gs_bit_3 << 13);
+    _function_data |= LOW_GS_ENHANCEMENT_MASK & new_data;
+}
+
+void TLC5957::setColorControl(uint16_t cc)
+{
+    setColorControl(cc, cc, cc);
+}
+
+void TLC5957::setColorControl(uint16_t ccr, uint16_t ccg, uint16_t ccb)
+{
+    uint64_t new_data = 0;
+
+    if (ccr > 511)
+        ccr = 511;
+    new_data |= ccr << 14;
+
+    if (ccg > 511)
+        ccg = 511;
+    new_data |= (uint32_t)ccg << 23;
+
+    if (ccb > 511)
+        ccb = 511;
+    new_data |= (uint64_t)ccb << 32;
+
+    _function_data |= COLOR_CONTROL_MASK & new_data;
+    _CC[0] = ccr;
+    _CC[1] = ccg;
+    _CC[2] = ccb;
+}
+
+void TLC5957::getColorControl(uint16_t* colorControl)
+{
+    colorControl[0] = _function_data >> 32 & 511;
```

Comment:
```suggestion
    colorControl[0] = (_function_data >> 32) & 0x01FF;
```

I'm always scared with bitshifting operators. they arent in PEDMAS and to me always confuse me in terms of order of precedence.